rust/basic_bitcoin/src/basic_bitcoin/src/bitcoin_wallet.rs

```diff
@@ -0,0 +1,325 @@
+
+    // Build the transaction that sends `amount` to the destination address.
+    let transaction = build_transaction(
+        &own_public_key,
+        &own_address,
+        &own_utxos,
+        &dst_address,
+        amount,
+        fee_per_byte,
+    )
+    .await
+    .expect("Error building transaction.");
+
+    let tx_bytes = transaction.serialize();
+    print(&format!("Transaction to sign: {}", hex::encode(tx_bytes)));
+
+    // Sign the transaction.
+    let signed_transaction = sign_transaction(
+        &own_public_key,
+        &own_address,
+        transaction,
+        derivation_path,
+        ecdsa_api::sign_with_ecdsa,
+    )
+    .await;
+
+    let signed_transaction_bytes = signed_transaction.serialize();
+    print(&format!(
+        "Signed transaction: {}",
+        hex::encode(&signed_transaction_bytes)
+    ));
+
+    print("Sending transaction...");
+    bitcoin_api::send_transaction(network, signed_transaction_bytes).await;
+    print("Done");
+}
+
+// Builds a transaction to send the given `amount` of satoshis to the
+// destination address.
+async fn build_transaction(
+    own_public_key: &[u8],
+    own_address: &Address,
+    own_utxos: &[Utxo],
+    dst_address: &Address,
+    amount: u64,
+    fee_per_byte: MillisatoshiPerByte,
+) -> Result<Transaction, String> {
+    // We have a chicken-and-egg problem where we need to know the length
+    // of the transaction in order to compute its proper fee, but we need
+    // to know the proper fee in order to figure out the inputs needed for
+    // the transaction.
+    //
+    // We solve this problem iteratively. We start with a fee of zero, build
+    // and sign a transaction, see what it's size is, and then update then
+    // update the fee, rebuild the transaction, until the fee is set to the
+    // correct amount.
+    print("Building transaction...");
+    let mut total_fee = 0;
+    loop {
+        print(&format!("Trying fee: {}", total_fee));
+        let transaction =
+            build_transaction_helper(own_utxos, own_address, dst_address, amount, total_fee)
+                .expect("Error building transaction.");
+
+        // Sign the transaction. In this case, we only care about the size
+        // of the signed transaction, so we use a mock signer here for efficiency.
+        let signed_transaction = sign_transaction(
+            own_public_key,
+            own_address,
+            transaction.clone(),
+            vec![], // mock derivation path
+            mock_signer,
+        )
+        .await;
+
+        let signed_tx_bytes_len = signed_transaction.serialize().len() as u64;
+
+        if (signed_tx_bytes_len * fee_per_byte) / 1000 == total_fee {
+            print(&format!("Transaction built with fee {}.", total_fee));
+            return Ok(transaction);
+        } else {
+            total_fee = (signed_tx_bytes_len * fee_per_byte) / 1000;
+        }
+    }
+}
+
+fn build_transaction_helper(
```

Comment:
Given the large amount of "helper code", did you consider splitting the code into one module for the main functions and another module for the helper functions to keep things tidy?
I'm aware that this would require the reader to open up another file to examine a particular piece of source code but that might be preferable to having hundreds of lines of code in one module. What do you think?

Reply:
I understand your concern. It is a bit subjective what's better in terms of readability (small modules over several files vs one larger module). I am personally in the latter camp, where I don't mind a file being a few hundred lines long, provided that it's coherent, especially that I find myself bikeshedding a lot on the naming of all the different files in the former case.

So, if it's ok with you, I'll keep things in one file, but I am renaming this method to not be called "helper", so now theoretically speaking it's not a helper method anymore :)